### PR TITLE
Add cache support to Azure dynamic inventory

### PR DESCRIPTION
**Pull request by pmarques**
##### SUMMARY

Update Azure RM dynamic inventory to support cache

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

contrib/inventory/azure_rm.ini
contrib/inventory/azure_rm.py

##### ADDITIONAL INFORMATION

This updated is based on GCE dynamic inventory #18093

### CODE REVIEW

File: contrib/inventory/azure_rm.py [hunk 1]
```diff
@@ -498,7 +543,22 @@ def __init__(self):
...
+            self.get_inventory()
+            self.cache.write_to_cache(self._inventory)
+        else:
+            self.load_inventory_from_cache()
+            cache_used = True
+            self._inventory['_meta']['stats'] = {'use_cache': True}
+
+        self._inventory['_meta']['stats'] = {
+            'inventory_load_time': time() - start_inventory_time,
+            'cache_used': cache_used
```
Comment (ZhijunZhao) [inline contrib/inventory/azure_rm.py]: Isn't `cache_used ` duplicated with `use_cache`?

File: contrib/inventory/azure_rm.py [hunk 2]
```diff
@@ -812,6 +897,12 @@ def _load_settings(self):
                 except:
                     pass
 
+            if not config.has_section('cache'):
+                config.add_section('cache')
```
Comment (ZhijunZhao) [inline contrib/inventory/azure_rm.py:1015]: I think you can simply skip it if config file doesn't have `cache` section. I assume this is for backward compatibility.

File: contrib/inventory/azure_rm.py [hunk 3]
```diff
@@ -812,6 +897,12 @@ def _load_settings(self):
...
+
+            settings['cache_path'] = config.get('cache', 'cache_path')
+            settings['cache_max_age'] = config.getint('cache', 'cache_max_age')
+            settings['cache_filename'] = config.get('cache', 'cache_filename')
```
Comment (ZhijunZhao) [inline contrib/inventory/azure_rm.py]: For this `cache` section, why don't you use a similar dict as `AZURE_CONFIG_SETTINGS `?

File: contrib/inventory/azure_rm.py [hunk 4]
```diff
@@ -799,7 +880,11 @@ def _load_settings(self):
         config = None
         settings = None
         try:
-            config = cp.ConfigParser()
+            config = cp.ConfigParser(defaults={
+                'cache_filename': 'ansible-azure_rm.cache',
+                'cache_path': '~/.ansible/tmp',
+                'cache_max_age': '300'
+            })
```
Comment (ZhijunZhao) [inline contrib/inventory/azure_rm.py:1001]: I'm a bit confused. Is this adding cached related configs to `defaults` section?

There are two places setting default values for cache config. I prefer to do default setting in a single place. How about just do it in `CloudInventoryCache`? Besides, it seems the `cache_path` config is not consistent in these two places.

**Review by ZhijunZhao — CHANGES_REQUESTED** [review]

File: contrib/inventory/azure_rm.py (hunk 3 above)
Comment (pmarques) [reply contrib/inventory/azure_rm.py]: I need to write a lot of code in a couple of places to accommodate that and tbh I don't the env vars are doing their job, they should overwrite what is read from a file and they are not doing that :(

File: contrib/inventory/azure_rm.py (hunk 2 above)
Comment (pmarques) [reply contrib/inventory/azure_rm.py:1015]: We get an error if the section doesn't exist. The original code is doing a *try catch*, what I don't like

File: contrib/inventory/azure_rm.py (hunk 4 above)
Comment (pmarques) [reply contrib/inventory/azure_rm.py:1001]: These are default values for Azure, CloudInventoryCache is more a generic thing and hopefully can be moved to Ansible utils or something that we can share among dynamic inventory scripts (at least written in python).
The **CloudInventoryCache** is a copy paste from gce.py and I would like to not change it now.

File: contrib/inventory/azure_rm.py (hunk 1 above)
Comment (pmarques) [reply contrib/inventory/azure_rm.py]: Ups, I copy pasted the version in gce.py :/

**Comment by yuwzho** [comment]
shipit

**Comment by jstoja** [comment]
@bcoca Would that be merge-able? Looks like everything's okay.

**Comment by yungezz** [comment]
@pmarques  could you pls resolve comments? is there any tests done for the change?

File: contrib/inventory/azure_rm.py (hunk 3 above)
Comment (yungezz) [reply contrib/inventory/azure_rm.py]: pls add those in environment variables for consistent configuration behavior

**Comment by pmarques** [comment]
@yungezz I applied the same patch to code that I used in several projects and not issues so far. Let me know if you had some.

**Comment by yungezz** [comment]
@pmarques pls add config : cache_path, cache_filename, cache_max_age as environment variables for consistent config behavior in the inventory.

**Comment by Fred-sun** [comment]
Hi @pmarques what do you think of @yungezz's option? Could you please help update the PR? Thanks!

**Comment by pmarques** [comment]
Updated.

Used inline env var lookup (with precedence) since **cache**  is In a different section. Also, I need this to be mixed with the config file which doesn't seem to be possible with other env vars

**Comment by Fred-sun** [comment]
@pmargara Thanks for your contribution, When you finished change, I will push maintainers to review. Thanks!

**Comment by pmarques** [comment]
@Fred-sun , yes

needs_revision

**Comment by Fred-sun** [comment]
@pmarques Thanks for you follow up and contribution.

**Comment by Fred-sun** [comment]
@pmarques Thanks for the contribute, Does this PR ready for review? If yes, I will push maintainers to review? Thanks!

**Comment by Fred-sun** [comment]
@pmarques Thanks for the contribution, Could help confirm the PR ready for review or not? If ready, I will push to review and merge. Thanks!

**Comment by Fred-sun** [comment]
@pmarques Thanks for the contribution, Could help confirm the PR ready for review or not? If ready, I will push to review and merge. Thanks!

**Comment by pmarques** [comment]
@yuwzho @yungezz any idea how to fix the bot?

**Comment by Fred-sun** [comment]
@yungezz @yuwzho Could you help take a look this PR? How about this PR going? merged or waiting further reviewed ? Thanks!

**Comment by Fred-sun** [comment]
@pmarques Thanks for your contribution. This inventory script is in maintenance mode, only critical bug fixes but no new features. There's new Azure external inventory script at [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/azure_rm.py) with better performance and latest new features. Pls move your contribution there. Thanks !

**Comment by pmarques** [comment]
@Fred-sun I'm already using that one for new projects, I also implemented some cache there: https://github.com/ansible/ansible/pull/48686

Do you want to close this one?

**Comment by Fred-sun** [comment]
yeah, close this and pay more attention on new. Thanks!
